db/db_pebble.go

```diff
@@ -0,0 +1,278 @@
+		}
+		return nil, err
+	}
+	val := make([]byte, len(v))
+	copy(val, v)
+	return val, closer.Close()
+}
+
+// Put inserts a <key, value> record
+func (b *PebbleDB) Put(ns string, key, value []byte) (err error) {
+	if !b.IsReady() {
+		return ErrDBNotStarted
+	}
+	err = b.db.Set(nsKey(ns, key), value, nil)
+	if err != nil {
+		if errors.Is(err, syscall.ENOSPC) {
+			log.L().Fatal("Failed to put db.", zap.Error(err))
+		}
+		err = errors.Wrap(ErrIO, err.Error())
+	}
+	return
+}
+
+// Delete deletes a record,if key is nil,this will delete the whole bucket
+func (b *PebbleDB) Delete(ns string, key []byte) (err error) {
+	if !b.IsReady() {
+		return ErrDBNotStarted
+	}
+	if key == nil {
+		panic("delete whole ns not supported by PebbleDB")
+	}
+	err = b.db.Delete(nsKey(ns, key), nil)
+	if err != nil {
+		if errors.Is(err, syscall.ENOSPC) {
+			log.L().Fatal("Failed to delete db.", zap.Error(err))
+		}
+		err = errors.Wrap(ErrIO, err.Error())
+	}
+	return
+}
+
+// WriteBatch commits a batch
+func (b *PebbleDB) WriteBatch(kvsb batch.KVStoreBatch) error {
+	if !b.IsReady() {
+		return ErrDBNotStarted
+	}
+
+	batch, err := b.dedup(kvsb)
+	if err != nil {
+		return nil
+	}
+	err = batch.Commit(nil)
+	if err != nil {
+		if errors.Is(err, syscall.ENOSPC) {
+			log.L().Fatal("Failed to write batch db.", zap.Error(err))
+		}
+		err = errors.Wrap(ErrIO, err.Error())
+	}
+	return err
+}
+
+func (b *PebbleDB) dedup(kvsb batch.KVStoreBatch) (*pebble.Batch, error) {
+	kvsb.Lock()
+	defer kvsb.Unlock()
+
+	type doubleKey struct {
+		ns  string
+		key string
+	}
+	// remove duplicate keys, only keep the last write for each key
+	var (
+		entryKeySet = make(map[doubleKey]struct{})
+		ch          = b.db.NewBatch()
+	)
+	for i := kvsb.Size() - 1; i >= 0; i-- {
+		write, e := kvsb.Entry(i)
+		if e != nil {
+			return nil, e
+		}
+		// only handle Put and Delete
+		if write.WriteType() != batch.Put && write.WriteType() != batch.Delete {
+			continue
+		}
+		key := write.Key()
+		k := doubleKey{ns: write.Namespace(), key: string(key)}
+		if _, ok := entryKeySet[k]; !ok {
+			entryKeySet[k] = struct{}{}
+			// add into batch
+			if write.WriteType() == batch.Put {
+				ch.Set(nsKey(write.Namespace(), key), write.Value(), nil)
+			} else {
+				ch.Delete(nsKey(write.Namespace(), key), nil)
+			}
+		}
+	}
+	return ch, nil
+}
+
+// Filter returns <k, v> pair in a bucket that meet the condition
+func (b *PebbleDB) Filter(ns string, cond Condition, minKey []byte, maxKey []byte) (keys [][]byte, vals [][]byte, err error) {
+	if !b.IsReady() {
+		return nil, nil, ErrDBNotStarted
+	}
+
+	iter, err := b.db.NewIter(&pebble.IterOptions{})
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to create iterator")
+	}
+	defer func() {
+		if e := iter.Close(); e != nil {
+			log.L().Error("Failed to close iterator", zap.Error(e))
+		}
+	}()
+	for iter.SeekPrefixGE(nsKey(ns, minKey)); iter.Valid(); iter.Next() {
```

Comment:
would `SeekPrefixGE()` return true for keys in another namespace with `nsKey() > nsKey(ns, minKey)`, for example, say:
nsKey(ns, minKey) = 0x12345678minkey ...
if another key in namespace `ns1` has:
nsKey(ns, key) = 0x12345679minkey ...

does `SeekPrefixGE` return true?

Reply:
no, `SeekPrefixGE` only returns values with same prefix, which is defined by `prefixLength`